File: LeetCode/337.House_Robber_III/337.House_Robber_III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] dp = helper(root, new int[2]);
        return Math.max(dp[0], dp[1]);
    }

    // dp[0]: not rob, dp[1]: rob
    private int[] helper(TreeNode node, int[] dp) {
        if (node == null) {
            return new int[2];
        }

        int[] left = helper(node.left, dp);
        int[] right = helper(node.right, dp);

        int[] next = new int[2];
        next[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        next[1] = node.val + left[0] + right[0];

        return next;
    }
}

File: LeetCode/337.House_Robber_III/337.House_Robber_III.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        # dfs return two state, one is not rob, one is rob

        def dfs(node):
            if not node:
                return (0, 0)  # rob, not rob

            left = dfs(node.left)
            right = dfs(node.right)

            # rob
            rob = left[1] + node.val + right[1]
            # not rob
            not_rob = max(left) + max(right)

            return (rob, not_rob)

        return max(dfs(root))
